File: crates/swss-common/src/types/configdbconnector.rs
use super::*;
use crate::bindings::*;
use std::collections::HashMap;

/// Rust wrapper around `swss::ConfigDBConnector_Native`.
#[derive(Debug)]
pub struct ConfigDBConnector {
    pub(crate) ptr: SWSSConfigDBConnector,
    use_unix_socket_path: bool,
    netns: String,
}

impl ConfigDBConnector {
    /// Create a new ConfigDBConnector.
    pub fn new(use_unix_socket_path: bool, netns: Option<String>) -> Result<ConfigDBConnector> {
        let netns_str = netns.unwrap_or_default();
        let netns_cstr = cstr(&netns_str)?;

        let ptr = unsafe {
            swss_try!(p_config_db => SWSSConfigDBConnector_new(
                use_unix_socket_path as u8,
                netns_cstr.as_ptr(),
                p_config_db
            ))?
        };

        Ok(Self {
            ptr,
            use_unix_socket_path,
            netns: netns_str,
        })
    }

    /// Connect to ConfigDB.
    ///
    /// # Arguments
    /// * `wait_for_init` - Wait for CONFIG_DB_INITIALIZED flag if true
    /// * `retry_on` - Retry connection on failure if true
    pub fn connect(&self, wait_for_init: bool, retry_on: bool) -> Result<()> {
        unsafe {
            swss_try!(SWSSConfigDBConnector_connect(
                self.ptr,
                wait_for_init as u8,
                retry_on as u8
            ))
        }
    }

    /// Get a single entry from a table.
    pub fn get_entry(&self, table: &str, key: &str) -> Result<HashMap<String, CxxString>> {
        let table_cstr = cstr(table)?;
        let key_cstr = cstr(key)?;

        unsafe {
            let arr = swss_try!(p_arr => SWSSConfigDBConnector_get_entry(
                self.ptr,
                table_cstr.as_ptr(),
                key_cstr.as_ptr(),
                p_arr
            ))?;
            take_field_value_array(arr)
        }
    }

    /// Get all keys from a table.
    pub fn get_keys(&self, table: &str, split: bool) -> Result<Vec<String>> {
        let table_cstr = cstr(table)?;

        unsafe {
            let arr = swss_try!(p_arr => SWSSConfigDBConnector_get_keys(
                self.ptr,
                table_cstr.as_ptr(),
                split as u8,
                p_arr
            ))?;
            take_string_array(arr)
        }
    }

    /// Get entire table as key-value pairs.
    /// Returns a HashMap where keys are table keys and values are field-value maps.
    pub fn get_table(&self, table: &str) -> Result<HashMap<String, HashMap<String, CxxString>>> {
        let table_cstr = cstr(table)?;

        unsafe {
            let arr = swss_try!(p_arr => SWSSConfigDBConnector_get_table(
                self.ptr,
                table_cstr.as_ptr(),
                p_arr
            ))?;

            let kfvs = take_key_op_field_values_array(arr)?;
            let mut table_map = HashMap::new();

            for kfv in kfvs {
                table_map.insert(kfv.key, kfv.field_values);
            }

            Ok(table_map)
        }
    }

    /// Set an entry in a table.
    pub fn set_entry<I, F, V>(&self, table: &str, key: &str, data: I) -> Result<()>
    where
        I: IntoIterator<Item = (F, V)>,
        F: AsRef<[u8]>,
        V: Into<CxxString>,
    {
        let table_cstr = cstr(table)?;
        let key_cstr = cstr(key)?;
        let (fv_array, _keep_alive) = make_field_value_array(data)?;

        unsafe {
            swss_try!(SWSSConfigDBConnector_set_entry(
                self.ptr,
                table_cstr.as_ptr(),
                key_cstr.as_ptr(),
                &fv_array
            ))
        }
    }

    /// Modify an entry in a table (update existing fields, add new ones).
    pub fn mod_entry<I, F, V>(&self, table: &str, key: &str, data: I) -> Result<()>
    where
        I: IntoIterator<Item = (F, V)>,
        F: AsRef<[u8]>,
        V: Into<CxxString>,
    {
        let table_cstr = cstr(table)?;
        let key_cstr = cstr(key)?;
        let (fv_array, _keep_alive) = make_field_value_array(data)?;

        unsafe {
            swss_try!(SWSSConfigDBConnector_mod_entry(
                self.ptr,
                table_cstr.as_ptr(),
                key_cstr.as_ptr(),
                &fv_array
            ))
        }
    }

    /// Delete an entire table.
    pub fn delete_table(&self, table: &str) -> Result<()> {
        let table_cstr = cstr(table)?;

        unsafe {
            swss_try!(SWSSConfigDBConnector_delete_table(
                self.ptr,
                table_cstr.as_ptr()
            ))
        }
    }

    /// Get connection info.
    pub fn use_unix_socket_path(&self) -> bool {
        self.use_unix_socket_path
    }

    /// Get netns.
    pub fn netns(&self) -> &str {
        &self.netns
    }

    /// Get the Redis client for the specified database.
    /// This allows direct Redis operations on the underlying database.
    /// Returns a borrowed reference to the underlying Redis client.
    pub fn get_redis_client(&self, db_name: &str) -> Result<BorrowedDbConnector> {
        let db_name_cstr = cstr(db_name)?;

        unsafe {
            let db_connector_ptr = swss_try!(p_db => SWSSConfigDBConnector_get_redis_client(
                self.ptr,
                db_name_cstr.as_ptr(),
                p_db
            ))?;

            // Create a borrowed wrapper around the returned pointer
            // Note: This is a reference to an existing connector, so we don't own it
            Ok(BorrowedDbConnector { ptr: db_connector_ptr })
        }
    }

    /// Get the entire configuration as a nested map structure.
    /// Returns a HashMap where keys are table names and values are table contents
    /// (key -> field -> value mappings).
    pub fn get_config(&self) -> Result<HashMap<String, HashMap<String, HashMap<String, CxxString>>>> {
        unsafe {
            let config_map = swss_try!(p_config => SWSSConfigDBConnector_get_config(
                self.ptr,
                p_config
            ))?;

            let mut result = HashMap::new();

            // Process each table
            for i in 0..config_map.len {
                let i_usize = i as usize;
                let table = &*config_map.data.add(i_usize);
                let table_name = std::ffi::CStr::from_ptr(table.table_name).to_string_lossy().to_string();

                let mut table_entries = HashMap::new();

                // Process each key in the table
                for j in 0..table.entries.len {
                    let j_usize = j as usize;
                    let entry = &*table.entries.data.add(j_usize);
                    let key = std::ffi::CStr::from_ptr(entry.key).to_string_lossy().to_string();

                    let mut field_values = HashMap::new();

                    // Process each field-value pair
                    for k in 0..entry.fieldValues.len {
                        let k_usize = k as usize;
                        let fv = &*entry.fieldValues.data.add(k_usize);
                        let field = std::ffi::CStr::from_ptr(fv.field).to_string_lossy().to_string();
                        // Note: We need to take ownership of the string, but fv.value will be freed
                        // by SWSSConfigMap_free, so we need to clone the string content first
                        let value_str = std::ffi::CStr::from_ptr(SWSSStrRef_c_str(fv.value as SWSSStrRef)).to_string_lossy().to_string();
                        let value = CxxString::new(value_str);

                        field_values.insert(field, value);
                    }

                    table_entries.insert(key, field_values);
                }

                result.insert(table_name, table_entries);
            }

            // Free the C structure
            SWSSConfigMap_free(config_map);

            Ok(result)
        }
    }
}

impl Drop for ConfigDBConnector {
    fn drop(&mut self) {
        // Ignore errors in Drop to avoid panic-during-panic scenarios
        let _ = unsafe { swss_try!(SWSSConfigDBConnector_free(self.ptr)) };
    }
}

unsafe impl Send for ConfigDBConnector {}

#[cfg(feature = "async")]
impl ConfigDBConnector {
    async_util::impl_basic_async_method!(new_async <= new(use_unix_socket_path: bool, netns: Option<String>) -> Result<ConfigDBConnector>);
    async_util::impl_basic_async_method!(connect_async <= connect(&self, wait_for_init: bool, retry_on: bool) -> Result<()>);
    async_util::impl_basic_async_method!(get_entry_async <= get_entry(&self, table: &str, key: &str) -> Result<HashMap<String, CxxString>>);
    async_util::impl_basic_async_method!(get_keys_async <= get_keys(&self, table: &str, split: bool) -> Result<Vec<String>>);
    async_util::impl_basic_async_method!(get_table_async <= get_table(&self, table: &str) -> Result<HashMap<String, HashMap<String, CxxString>>>);
    async_util::impl_basic_async_method!(get_config_async <= get_config(&self) -> Result<HashMap<String, HashMap<String, HashMap<String, CxxString>>>>);
    async_util::impl_basic_async_method!(delete_table_async <= delete_table(&self, table: &str) -> Result<()>);
}

/// A borrowed reference to a DbConnector obtained from ConfigDBConnector.
/// This does not own the underlying connection and does not implement Drop.
#[derive(Debug)]
pub struct BorrowedDbConnector {
    ptr: SWSSDBConnector,
}

impl BorrowedDbConnector {
    /// Create a new BorrowedDbConnector from a raw pointer.
    /// This is intended for internal use by other modules.
    pub(crate) fn new(ptr: SWSSDBConnector) -> Self {
        Self { ptr }
    }

    /// Flush the current database, removing all keys.
    /// Equivalent to Redis FLUSHDB command.
    pub fn flush_db(&self) -> Result<bool> {
        let status = unsafe { swss_try!(p_status => SWSSDBConnector_flushdb(self.ptr, p_status))? };
        Ok(status == 1)
    }
}

File: common/c-api/configdbconnector.cpp
#include <cstring>
#include <string>
#include <utility>
#include <map>
#include <vector>

#include "../configdb.h"
#include "configdbconnector.h"
#include "util.h"

using namespace swss;
using namespace std;

SWSSResult SWSSConfigDBConnector_new(uint8_t use_unix_socket_path, const char *netns, SWSSConfigDBConnector *outConfigDb) {
    SWSSTry({
        string netns_str = netns ? string(netns) : "";
        *outConfigDb = (SWSSConfigDBConnector) new ConfigDBConnector_Native(use_unix_socket_path != 0, netns_str.c_str());
    });
}

SWSSResult SWSSConfigDBConnector_free(SWSSConfigDBConnector configDb) {
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wdelete-non-virtual-dtor"
    SWSSTry(delete (ConfigDBConnector_Native *)configDb);
#pragma GCC diagnostic pop
}

SWSSResult SWSSConfigDBConnector_connect(SWSSConfigDBConnector configDb, uint8_t wait_for_init, uint8_t retry_on) {
    SWSSTry(((ConfigDBConnector_Native *)configDb)->connect(wait_for_init != 0, retry_on != 0));
}

SWSSResult SWSSConfigDBConnector_get_entry(SWSSConfigDBConnector configDb, const char *table, const char *key, SWSSFieldValueArray *outEntry) {
    SWSSTry({
        auto entry_map = ((ConfigDBConnector_Native *)configDb)->get_entry(string(table), string(key));
        // Convert map<string, string> to vector<pair<string, string>> for makeFieldValueArray
        vector<pair<string, string>> pairs;
        pairs.reserve(entry_map.size());
        for (auto &pair : entry_map) {
            pairs.push_back(make_pair(pair.first, move(pair.second)));
        }

        *outEntry = makeFieldValueArray(std::move(pairs));
    });
}

SWSSResult SWSSConfigDBConnector_get_keys(SWSSConfigDBConnector configDb, const char *table, uint8_t split, SWSSStringArray *outKeys) {
    SWSSTry({
        auto keys = ((ConfigDBConnector_Native *)configDb)->get_keys(string(table), split != 0);
        *outKeys = makeStringArray(std::move(keys));
    });
}

SWSSResult SWSSConfigDBConnector_get_table(SWSSConfigDBConnector configDb, const char *table, SWSSKeyOpFieldValuesArray *outTable) {
    SWSSTry({
        auto table_map = ((ConfigDBConnector_Native *)configDb)->get_table(string(table));

        // Convert map<string, map<string, string>> to vector<SWSSKeyOpFieldValues>
        vector<SWSSKeyOpFieldValues> table_entries;
        table_entries.reserve(table_map.size());

        for (auto &entry : table_map) {
            SWSSKeyOpFieldValues kfv_entry;
            kfv_entry.key = strdup(entry.first.c_str());
            kfv_entry.operation = SWSSKeyOperation_SET; // ConfigDB entries are always SET operations

            // Convert inner map to field-value array
            vector<pair<string, string>> pairs;
            pairs.reserve(entry.second.size());
            for (auto &field_pair : entry.second) {
                pairs.push_back(make_pair(field_pair.first, move(field_pair.second)));
            }
            kfv_entry.fieldValues = makeFieldValueArray(std::move(pairs));

            table_entries.push_back(kfv_entry);
        }

        // Convert to SWSSKeyOpFieldValuesArray
        SWSSKeyOpFieldValues *data = new SWSSKeyOpFieldValues[table_entries.size()];
        for (size_t i = 0; i < table_entries.size(); i++) {
            data[i] = table_entries[i];
        }

        SWSSKeyOpFieldValuesArray out;
        out.len = (uint64_t)table_entries.size();
        out.data = data;
        *outTable = out;
    });
}

SWSSResult SWSSConfigDBConnector_set_entry(SWSSConfigDBConnector configDb, const char *table, const char *key, const SWSSFieldValueArray *data) {
    SWSSTry({
        // Convert SWSSFieldValueArray to map<string, string>
        map<string, string> data_map;
        auto field_values = takeFieldValueArray(*data);
        for (auto &fv : field_values) {
            data_map[fv.first] = fv.second;
        }

        ((ConfigDBConnector_Native *)configDb)->set_entry(string(table), string(key), data_map);
    });
}

SWSSResult SWSSConfigDBConnector_mod_entry(SWSSConfigDBConnector configDb, const char *table, const char *key, const SWSSFieldValueArray *data) {
    SWSSTry({
        // Convert SWSSFieldValueArray to map<string, string>
        map<string, string> data_map;
        auto field_values = takeFieldValueArray(*data);
        for (auto &fv : field_values) {
            data_map[fv.first] = fv.second;
        }

        ((ConfigDBConnector_Native *)configDb)->mod_entry(string(table), string(key), data_map);
    });
}

SWSSResult SWSSConfigDBConnector_delete_table(SWSSConfigDBConnector configDb, const char *table) {
    SWSSTry(((ConfigDBConnector_Native *)configDb)->delete_table(string(table)));
}

SWSSResult SWSSConfigDBConnector_get_redis_client(SWSSConfigDBConnector configDb, const char *db_name, SWSSDBConnector *outDbConnector) {
    SWSSTry({
        // Get the Redis client from the ConfigDBConnector's parent SonicV2Connector_Native
        DBConnector& redis_client = ((ConfigDBConnector_Native *)configDb)->get_redis_client(string(db_name));

        // Return a pointer to the existing DBConnector
        // Note: This returns a reference to the existing connector, not a new one
        *outDbConnector = (SWSSDBConnector)&redis_client;
    });
}

SWSSResult SWSSConfigDBConnector_get_config(SWSSConfigDBConnector configDb, SWSSConfigMap *outConfig) {
    SWSSTry({
        // Get the entire configuration as a nested map structure
        auto config_map = ((ConfigDBConnector_Native *)configDb)->get_config();

        // Convert map<string, map<string, map<string, string>>> to SWSSConfigMap
        vector<SWSSConfigTable> config_tables;
        config_tables.reserve(config_map.size());

        for (auto &table_entry : config_map) {
            SWSSConfigTable config_table;
            config_table.table_name = strdup(table_entry.first.c_str());

            // Convert map<string, map<string, string>> to SWSSKeyOpFieldValuesArray
            vector<SWSSKeyOpFieldValues> table_entries;
            table_entries.reserve(table_entry.second.size());

            for (auto &key_entry : table_entry.second) {
                SWSSKeyOpFieldValues kfv_entry;
                kfv_entry.key = strdup(key_entry.first.c_str());
                kfv_entry.operation = SWSSKeyOperation_SET; // Config entries are always SET operations

                // Convert map<string, string> to SWSSFieldValueArray
                vector<pair<string, string>> pairs;
                pairs.reserve(key_entry.second.size());
                for (auto &field_pair : key_entry.second) {
                    pairs.push_back(make_pair(field_pair.first, move(field_pair.second)));
                }
                kfv_entry.fieldValues = makeFieldValueArray(std::move(pairs));

                table_entries.push_back(kfv_entry);
            }

            // Convert vector to array
            SWSSKeyOpFieldValues *data = new SWSSKeyOpFieldValues[table_entries.size()];
            for (size_t i = 0; i < table_entries.size(); i++) {
                data[i] = table_entries[i];
            }

            config_table.entries.len = (uint64_t)table_entries.size();
            config_table.entries.data = data;

            config_tables.push_back(config_table);
        }

        // Convert vector to final array
        SWSSConfigTable *config_data = new SWSSConfigTable[config_tables.size()];
        for (size_t i = 0; i < config_tables.size(); i++) {
            config_data[i] = config_tables[i];
        }

        outConfig->len = (uint64_t)config_tables.size();
        outConfig->data = config_data;
    });
}
